pkg/controller/grouping/convert.go

```diff
@@ -0,0 +1,69 @@
+// Copyright 2021 Antrea Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package grouping
+
+import (
+	"fmt"
+
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/klog"
+)
+
+func statusErrorWithMessage(msg string, params ...interface{}) metav1.Status {
+	return metav1.Status{
+		Message: fmt.Sprintf(msg, params...),
+		Status:  metav1.StatusFailure,
+	}
+}
+
+func ConvertClusterGroupCRD(Object *unstructured.Unstructured, toVersion string) (*unstructured.Unstructured, metav1.Status) {
+	klog.V(2).Infof("Converting CRD for ClusterGroup %s", Object.GetName())
+	convertedObject := Object.DeepCopy()
+	fromVersion := Object.GetAPIVersion()
+	if toVersion == fromVersion {
+		return nil, statusErrorWithMessage("conversion from a version to itself should not call the webhook: %s", toVersion)
+	}
+	switch Object.GetAPIVersion() {
+	case "crd.antrea.io/v1alpha2":
+		switch toVersion {
+		case "crd.antrea.io/v1alpha3":
+			ipb, found, err := unstructured.NestedMap(convertedObject.Object, "spec", "ipBlock")
+			if err == nil && found {
```

Comment:
Just to confirm it's not possible that the ipBlock field is found but empty

Reply:
After trying it out it seems it is actually possible... i.e. the following was a valid spec
```
apiVersion: crd.antrea.io/v1alpha2
kind: ClusterGroup
metadata:
  name: ipb-cg
spec:
  ipBlock:
```
However this will not cause any issues for the conversion logic per se. This spec can successfully be converted to v1alpha3 w/o problem, and the effect is this CG selects nothing. Since we cannot change how validation was done in the past, I would say this might better be categorized as a separate issue